File: iteration2/src/NotificationsInterface.java
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Scanner;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

public class NotificationsInterface {
    Colors Colors = new Colors();
    private ArrayList<Notification> notifications = new ArrayList<>();
    private Scanner input = new Scanner(System.in);
    private Person person;

    public NotificationsInterface(Session session) {
        this.person = session.getUser();
        notifications.clear();
        calculateNotifications();

    }

    // shows the notifications
    private void showNotifications() {
        // indexer is for showing notifications with numbers or not

        if (notifications.isEmpty()) {
            System.out.println(
                    Colors.getYELLOW() + "There are no notifications available at the moment!" + Colors.getRESET());
            System.out.println(
                    "You will be " + Colors.getGREEN() + "directed" + Colors.getRESET() + " to the Main Menu!");
            return;
        }

        int i = 1;
        String Color;

        for (Notification notification : notifications) {

            String isRead = notification.getIsRead() ? "Read" : "Unread";
            String senderID = notification.getSenderID();
            String timeSent = notification.getTimeSent().toString();
            String description = notification.getDescription();
            String timeDate;
            String time;
            String isNew;

            if (isRead.equals("Read")) {
                Color = Colors.getBLUE();
                isNew = "";
            } else {
                Color = Colors.getGREEN();
                isNew = "New";
            }

            // System.out.println("--------------------");
            // System.out.printf("%n%s",BG);

            // System.out.printf("o %-15s%n\t%-15s%n\t%30s", senderID, description,
            // timeSent);
            System.out.println(i + "  " + Color + isNew + " Notification " + Colors.getRESET() + "\t" + timeSent);
            System.out.println("\tDescription: ");
            System.out.println(Colors.getBLUE() + "\t  o  " + Colors.getRESET() + Colors.getITALIC() + description
                    + Colors.getRESET());
            System.out.print(Colors.getRESET());
            System.out.println("");

            i++;
        }
    }

    // calculates the notifications by reading json file
    private void calculateNotifications() {
        try {
            Object notificationObject = new JSONParser().parse(new FileReader("./jsons/notifications.json"));
            JSONArray notificationJSONObject = (JSONArray) notificationObject;

            for (Object notification : notificationJSONObject) {
                JSONObject notificationJSON = (JSONObject) notification;
                if (notificationJSON.get("receiverID").equals(person.getID())) {

                    String receiverID = (String) notificationJSON.get("receiverID");
                    String description = (String) notificationJSON.get("description");
                    String senderID = (String) notificationJSON.get("senderID");
                    boolean isRead = (boolean) notificationJSON.get("isRead");
                    long notificationID = (long) notificationJSON.get("notificationID");

                    Notification newNotification = new Notification(receiverID, description, senderID, isRead,
                            notificationID);
                    notifications.add(newNotification);
                }
            }

        } catch (Exception e) {
            System.out.println(Colors.getYELLOW() + "Error: " + Colors.getRESET() + e);
        }
    }

    // shows the notifications menu and let user take action in notifications
    public void notificationsMenu() {

        // showNotifications();
        // if (notifications.isEmpty()) {
        // return;
        // }

        while (true) {
            System.out
                    .println(Colors.getRED() + Colors.getBOLD() + "\n>> Notifications Menu\n" + Colors.getRESET());

            // notifications
            showNotifications();
            if (notifications.isEmpty()) {
                return;
            }

            System.out.println(Colors.getYELLOW() + "1" + Colors.getRESET() + ".   Mark as Read");
            System.out.println(Colors.getYELLOW() + "2" + Colors.getRESET() + ".   Delete Notification");
            System.out.println(Colors.getYELLOW() + "0" + Colors.getRESET() + ".   Go back to the Main Menu");
            System.out.print("\n" + Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");
            System.out.print(Colors.getBLUE());
            char choice = input.next().charAt(0);
            System.out.print(Colors.getRESET());
            switch (choice) {

                // mark as read
                case '1':
                    markAsRead();
                    break;

                // delete notification
                case '2':
                    deleteNotification();
                    break;

                // go back to student menu
                case '0':
                    return;

                // invalid input
                default:
                    System.out.println(Colors.getYELLOW() + "Invalid input! Please try again." + Colors.getRESET());
                    break;
            }
        }

    }

    private void markAsRead() {
        // notifications with numbers

        while (true) {
            System.out.println(Colors.getRED() + Colors.getBOLD() + "\n>>> Mark as Read Menu\n" + Colors.getRESET()
                    + Colors.getRESET());
            showNotifications();

            System.out.println("Select a notification " + Colors.getBLUE() + "number" + Colors.getRESET()
                    + " to mark as read or select " + Colors.getBLUE() + "x" + Colors.getRESET()
                    + " to mark all as read.");
            System.out.println(
                    "(For example -> " + Colors.getYELLOW() + "1-2-3" + Colors.getRESET() + " or" + Colors.getYELLOW()
                            + "  x" + Colors.getRESET() + " )");
            System.out.println("Which notification do you want to mark as read? \n");
            System.out
                    .println(Colors.getYELLOW() + "0" + Colors.getRESET() + ".  Go back to the Notifications Menu.\n");

            System.out.print(Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");
            System.out.print(Colors.getBLUE());
            char choice = input.next().charAt(0);
            System.out.print(Colors.getRESET());
            switch (choice) {

                // go back to notifications menu
                case '0':
                    return;

                // mark all as read
                case 'x':
                    for (Notification notification : notifications) {
                        notification.setIsRead(true);
                    }
                    break;

                // mark as read
                default:
                    notifications.get(choice - 49).setIsRead(true);
                    break;
            }

        }

    }

    private void deleteNotification() {
        // notifications with numbers
        while (true) {
            System.out.println(Colors.getRED() + Colors.getBOLD() + "\n>>> Delete Notification Menu\n"
                    + Colors.getRESET() + Colors.getRESET());
            showNotifications();
            if (notifications.isEmpty()) {
                return;
            }
            System.out.println("Select a notification number to delete or select x to delete all.");

            System.out.println(
                    "(For example -> " + Colors.getYELLOW() + "1-2-3" + Colors.getRESET() + " or" + Colors.getYELLOW()
                            + "  x" + Colors.getRESET() + " )");
            System.out.println("Which notification do you want to delete? \n");
            System.out
                    .println(Colors.getYELLOW() + "0" + Colors.getRESET() + ".  Go back to the Notifications Menu.\n");
            System.out.print(Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");
            System.out.print(Colors.getBLUE());
            char choice = input.next().charAt(0);
            System.out.print(Colors.getRESET());

            switch (choice) {

                // go back to notifications menu
                case '0':
                    return;

                // delete all
                case 'x':
                    notifications.clear();
                    break;

                // delete
                default:
                    // delete from json
                    deleteInJSON(notifications.get(choice - 49).getNotificationID());
                    notifications.remove(choice - 49);
                    break;
            }

        }
    }

    private void deleteInJSON(long notificationID) {
        try {
            Object notificationObject = new JSONParser().parse(new FileReader("./jsons/notifications.json"));
            JSONArray notificationJSONObject = (JSONArray) notificationObject;

            for (Object notification : notificationJSONObject) {
                JSONObject notificationJSON = (JSONObject) notification;
                if (notificationJSON.get("notificationID").equals(notificationID)) {
                    notificationJSONObject.remove(notification);
                    PrintWriter pw = new PrintWriter("./jsons/notifications.json");
                    pw.write(notificationJSONObject.toJSONString());

                    pw.flush();
                    pw.close();
                }
            }

        } catch (Exception e) {
            System.out.println(Colors.getYELLOW() + "Error: " + Colors.getRESET() + e);
        }
    }

}

File: iteration3/src/notifications_interface.py
import json
from colors import Colors
from notification import Notification
from session import Session


class NotificationsInterface:
    def __init__(self, session: Session):
        self.__colors = Colors()
        self.__person = session.get_user()
        self.__notifications: list[Notification] = []
        self.__calculate_notifications()

    def __calculate_notifications(self):
        try:
            with open("jsons/notifications.json", "r") as notification_file:
                notification_object = json.load(notification_file)
            for notification_json in notification_object:
                if notification_json["receiverID"] == self.__person.get_id():
                    receiver_id = notification_json["receiverID"]
                    description = notification_json["description"]
                    sender_id = notification_json["senderID"]
                    is_read = notification_json["isRead"]
                    notification_id = notification_json["notificationID"]

                    new_notification = Notification(
                        receiver_id, description, sender_id, is_read, notification_id
                    )
                    self.__notifications.append(new_notification)

        except Exception as e:
            print(
                self.__colors.get_yellow()
                + "Error: "
                + str(e)
                + self.__colors.get_reset()
            )

    def __show_notifications(self):
        if not self.__notifications:
            print(
                self.__colors.get_yellow()
                + "You have no notifications."
                + self.__colors.get_reset()
            )
            print(
                "You will be "
                + self.__colors.get_green()
                + "directed"
                + self.__colors.get_reset()
                + " to the Main Menu!"
            )
            return
        for i, notification in enumerate(self.__notifications, 1):
            is_read = "Read" if notification.get_is_read() else "Unread"
            if is_read == "Read":
                color = self.__colors.get_blue()
                is_new = ""
            else:
                color = self.__colors.get_green()
                is_new = "New"

            print(
                f"{i}  {color}{is_new} Notification {self.__colors.get_reset()}\t{notification.get_time_sent()}"
            )
            print("\tDescription: ")
            print(
                f"{self.__colors.get_blue()}\t  o  {self.__colors.get_green()}{self.__colors.get_italic()}{notification.get_description()}{self.__colors.get_reset()}"
            )
            print(self.__colors.get_reset())

    def notifications_menu(self):
        while True:
            print(
                f"{self.__colors.get_red()}{self.__colors.get_bold()}\n>> Notifications Menu{self.__colors.get_reset()}"
            )

            # notifications
            self.__show_notifications()
            if not self.__notifications:
                return

            print(
                f"{self.__colors.get_yellow()}1{self.__colors.get_reset()}.   Mark as Read"
            )
            print(
                f"{self.__colors.get_yellow()}2{self.__colors.get_reset()}.   Delete Notification"
            )
            print(
                f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.   Go back to the Main Menu"
            )
            user_input = input(
                f"\n{self.__colors.get_blue()}--> {self.__colors.get_reset()}What do you want to do?   "
            )
            choice = user_input[0]

            switch_choice = {
                "1": lambda: self.__mark_as_read(),
                "2": lambda: self.__delete_notification(),
                "0": lambda: exit(),
                "default": lambda: print(
                    f"{self.__colors.get_yellow()}Invalid input! Please try again.{self.__colors.get_reset()}"
                ),
            }

            switch_choice.get(choice, switch_choice["default"])()

    def __mark_as_read(self):
        while True:
            print(
                f"{self.__colors.get_red()}{self.__colors.get_bold()}\n>> Mark as Read{self.__colors.get_reset()}"
            )
            self.__show_notifications()

            print(
                f"Select a notification{self.__colors.get_blue()} number {self.__colors.get_reset()}to mark as read or select {self.__colors.get_blue()}x{self.__colors.get_reset()} to mark all as read."
            )
            print(
                f"(For example -> {self.__colors.get_yellow()}1-2-3 {self.__colors.get_reset()}or {self.__colors.get_yellow()}x{self.__colors.get_reset()})"
            )
            print(
                f"\n{self.__colors.get_blue()}--> {self.__colors.get_reset()}Which notification do you want to mark as read?   "
            )
            print(
                f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.   Go back to the Notifications Menu"
            )

            user_input = input(
                f"{self.__colors.get_blue()}--> {self.__colors.get_reset()}What do you want to do?   "
            )
            choice = user_input[0]

            switch_choice = {
                "0": lambda: exit(),
                "x": lambda: self.__mark_all_as_read(),
                "default": lambda: self.__mark_single_as_read(int(choice)),
            }

            switch_choice.get(choice, switch_choice["default"])()

    def __mark_all_as_read(self):
        for notification in self.__notifications:
            notification.set_is_read(True)

    def __mark_single_as_read(self, index: int):
        self.__notifications[index - 1].set_is_read(True)

    def __delete_notification(self):
        # notifications with numbers
        while True:
            print(
                self.__colors.get_red()
                + self.__colors.get_bold()
                + "\n>>> Delete Notification Menu"
                + self.__colors.get_reset()
                + self.__colors.get_reset()
            )
            self.__show_notifications()
            if not self.__notifications:
                return

            print("Select a notification number to delete or select x to delete all.")
            print(
                f"(For example -> {self.__colors.get_yellow()}1{self.__colors.get_reset()} or{self.__colors.get_yellow()} x{self.__colors.get_reset()})"
            )
            print("Which notification do you want to delete? \n")
            print(
                f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.  Go back to the Notifications Menu.\n"
            )
            user_input = input(
                f"{self.__colors.get_blue()}--> {self.__colors.get_reset()}What do you want to do?   "
            )
            choice = user_input[0]

            # go back to notifications menu
            if choice == "0":
                return
            # delete all
            elif choice == "x":
                self.__notifications.clear()
                break
            # delete
            else:
                # delete from json
                self.__delete_in_json(
                    self.__notifications[int(choice) - 1].get_notification_id()
                )
                del self.__notifications[int(choice) - 1]
                break

    def __delete_in_json(self, notification_id):
        try:
            with open("./jsons/notifications.json", "r") as notification_file:
                notification_object: list = json.load(notification_file)

            for notification in notification_object:
                if notification["notificationID"] == notification_id:
                    notification_object.remove(notification)

                    with open("./jsons/notifications.json", "w") as notification_file:
                        json.dump(notification_object, notification_file, indent=2)

        except Exception as e:
            print(
                self.__colors.get_yellow()
                + "Error: "
                + self.__colors.get_reset()
                + str(e)
            )
